### ISSUE TITLE
Rule proposal: Require explicit undefined | never in optional object property types

### ISSUE THREAD

**Issue by pfgithub** [typescript-eslint/typescript-eslint#3794]
Typescript 4.4 adds `exactOptionalPropertyTypes` which changes all optional property types to no longer be implicitly `undefined | T` 

With this change, it is easy to write `{a?: T}` when you intended to write `{a?: undefined | T}`, requiring users of your code to do conditional object spreading when it may not be necessary. Additionally, it may be difficult to migrate to using exactOptionalPropertyTypes because you have to manually search for instances of `?:` to make sure they are all defined correctly.

This rule would require that all optional property types either be `undefined | T` or `never | T` to make it explicit if undefined should be allowed as a value or not.

Examples of **incorrect** code for this rule:

```ts
type Props = {
    onClick?: () => void,
};
```

Examples of **correct** code for this rule:

```ts
type Props = {
    onClick?: undefined | (() => void),
};

type Props = {
    // if {onClick: undefined} should be disallowed:
    onClick?: never | (() => void),
};
```

I wrote my own implementation of this [here](https://github.com/pfgithub/threadclient/blob/0bb1af6cc461b53978542d9a92f655796384470b/packages/eslint-plugin-custom-quote-rule/src/explicit-optional-key.js#L36) that can be used as a base if it is wanted in typescript-eslint

**Comment by bradzacher** [comment]
I do think that there is *some* value in this - but overall I think that the wider applicability of the lint rule is limited. I think most codebases aren't libraries with external consumers and are instead closed-systems - thus such patterns are spottable and fixable in a single change.

[Additionally this can *mostly* be covered by `no-restricted-syntax`](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&showAST=es&code=PQKgBApgzgNglgOwC5gQewLQCdpK3AYyQgBMMoBPZAQwA8AuMAbQCgx2wByCLLNLTgBo2HAN4iOHKBBgQi-RpwAqAZQAKfAA48kFFXADmCakgCuOJmk1I4aYzDABeMHlMQAumAB8YVUoraAIII6EgmtgjeYPShABSqAKoIEf7a9AAW1FDxKkkkEABmiKQA0hAUAO78JACUNUISkmAAttBQ1AYQilY2dtQOmlpQLaZQKKbJdmAABhP5RQik0w2SAL7CniDALCyIxFgF1AQQYABiaGhg4pI9Ef0aVgCMAPyMCKbNAEY8ANyNt30YA9NAAmV6oD7fLBgAA+YDG+AQBj+N2sdyBWgA6nAkOk8oVwe8vjxYWA5oViiQURwcABHUxwHAkYFvSG-FirFi6bRgABC1GhzmuHAB9mBL1ZxKw1PYovuWjBkqhpIRiGR-zRgOB2Nx+IKhLZ0Lh5IWpBlYDpDKZLIhUr+qyAA&rules=N4XyA&tsConfig=N4XyA). There will be some false-positives (eg if your property type is a type reference with `undefined` within it), but it'll cover the vast majority of cases.

For these reasons - we will be rejecting this proposal within this project.
Thanks for raising an issue!
